### PR TITLE
ci: test against more compilers, setup clippy and fix clippy lints

### PR DESCRIPTION
**Pull request by cpcloud**
- Use cargo flags to prevent out-of-date Cargo.lock
- Ship the Cargo.lock file in the source distribution
- ci: add more variations to ci

### CODE REVIEW

**Comment by Jimexist** [comment]
Thanks @cpcloud - i've incorporated some of your changes already, please rebase first?

**Comment by cpcloud** [comment]
@Jimexist done!

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -22,50 +22,66 @@ on:
...
     strategy:
       fail-fast: false
       matrix:
         python-version:
           - "3.7"
           - "3.10"
+        toolchain:
+          - stable
+          - beta
+          - nightly
```
[the remark this answers is not in the document]
Comment (Jimexist) [reply .github/workflows/test.yaml:41]: usually if you run against all of them there will be conflicting fmt changes

[the remark this answers is not in the document]
Comment (cpcloud) [reply .github/workflows/test.yaml:41]: The format check is only against stable

**Review by Jimexist — APPROVED** [review]

**Pull request merged**